### ISSUE TITLE
OpenSSL on Win10 x64 fails to build

### ISSUE THREAD

**Issue by alexander-novo** [openssl/openssl#1086]
`nmake` runs successfully, but then `nmake test` fails on just about every test, starting with rand. Unrelated to #1085 - I have tried building it in a folder without any spaces in the name.

<details>
<summary>`nmake test` output</summary>
```
        C:\Perl64\bin\perl.exe .\test\run_tests.pl
test\recipes\01-test_abort.t ............ ok
test\recipes\01-test_ordinals.t ......... ok
test\recipes\01-test_symbol_presence.t .. ok
test\recipes\05-test_bf.t ............... ok
test\recipes\05-test_cast.t ............. ok
test\recipes\05-test_des.t .............. ok
test\recipes\05-test_hmac.t ............. ok
test\recipes\05-test_idea.t ............. ok
test\recipes\05-test_md2.t .............. skipped: md2 is not supported by this OpenSSL build
test\recipes\05-test_md4.t .............. ok
test\recipes\05-test_md5.t .............. ok
test\recipes\05-test_mdc2.t ............. ok
test\recipes\05-test_rand.t ............. Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\05-test_rand.t ............. 1/1 #   Failed test 'running randtest'
#   at test\testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
test\recipes\05-test_rand.t ............. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
test\recipes\05-test_rc2.t .............. ok
test\recipes\05-test_rc4.t .............. ok
test\recipes\05-test_rc5.t .............. skipped: rc5 is not supported by this OpenSSL build
test\recipes\05-test_rmd.t .............. ok
test\recipes\05-test_sha1.t ............. ok
test\recipes\05-test_sha256.t ........... ok
test\recipes\05-test_sha512.t ........... ok
test\recipes\05-test_wp.t ............... ok
test\recipes\10-test_bn.t ............... 1/3 Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\10-test_bn.t ............... 2/3 #   Failed test 'initialize'
#   at test\recipes\10-test_bn.t line 24.
# Looks like you failed 1 test of 3.
test\recipes\10-test_bn.t ............... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/3 subtests
        (less 1 skipped subtest: 1 okay)
test\recipes\10-test_exp.t .............. Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\10-test_exp.t .............. 1/1 #   Failed test 'running exptest'
#   at test\testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
test\recipes\10-test_exp.t .............. Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
test\recipes\15-test_dh.t ............... Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\15-test_dh.t ............... 1/1 #   Failed test 'running dhtest'
#   at test\testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
test\recipes\15-test_dh.t ............... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
test\recipes\15-test_dsa.t .............. 1/6 Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\15-test_dsa.t .............. 2/6 #   Failed test 'running dsatest'
#   at test\recipes\15-test_dsa.t line 23.
Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

test\recipes\15-test_dsa.t .............. 3/6 #   Failed test 'running dsatest -app2_1'
#   at test\recipes\15-test_dsa.t line 24.
Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.

    #   Failed test 'p -> d'
    #   at ..\test\recipes\tconversion.pl line 60.
Terminating on signal SIGINT(2)
Terminating on signal SIGINT(2)
NMAKE : fatal error U1058: terminated by user
Stop.
NMAKE : fatal error U1077: 'C:\Perl64\bin\perl.exe' : return code '0x2'
Stop.
```
</details>
(cut off at end by me, but trust me - it would have failed just about every test afterwards)

**Comment by levitte** [comment]
Could you try this for me?

```
nmake VERBOSE=1 TESTS=test_rand
```

**Comment by alexander-novo** [comment]
I ran that command exactly, and it seemed to build just like normal, so I ran `nmake test VERBOSE=1 TESTS=test_rand` afterwards. It produced this:

```
        set SRCTOP=.
        set BLDTOP=.
        set PERL=C:\Perl64\bin\perl.exe
        C:\Perl64\bin\perl.exe .\test\run_tests.pl test_rand
test\recipes\05-test_rand.t ..
1..1
randtest.exe => 0
ok 1 - running randtest
ok
All tests successful.
Files=1, Tests=1, 31 wallclock secs ( 0.06 usr +  0.05 sys =  0.11 CPU)
Result: PASS
```

Even though it says "PASS", the test crashed. [Here's my entire log](https://paste.ee/p/Eqy2V).

**Comment by levitte** [comment]
So...  I'm getting increasingly unsure if this is a fault in OpenSSL.  From your description, it sounds like the fault can be a bit random, and all having to do with `cmd.exe` suddenly not being spawned properly.  Going back to your initial output, there are a number of tests at the start that run perfectly, and they do spawn `cmd.exe` just as much as the `rand` test does...

**Comment by alexander-novo** [comment]
And yet it's consistently the same tests which fail.

**Comment by levitte** [comment]
Ok, how about you try `nmake VERBOSE=1 test`, see if that gives a clue?  Warning, that's a LOT of output

**Comment by alexander-novo** [comment]
<details>
<summary> Output Log</summary>
```
C:\C++Libs\openssl>nmake VERBOSE=1 test

Microsoft (R) Program Maintenance Utility Version 14.00.23918.0
Copyright (C) Microsoft Corporation.  All rights reserved.

        set SRCTOP=.
        set BLDTOP=.
        set PERL=C:\Perl64\bin\perl.exe
        C:\Perl64\bin\perl.exe .\test\run_tests.pl
test\recipes\01-test_abort.t ............
1..1
test\aborttest.c:14: OpenSSL internal error: Voluntary abort
aborttest.exe => 3
ok 1 - Testing that abort is caught correctly
ok
test\recipes\01-test_ordinals.t .........
1..2
ok 1 - Test libcrypto.num
ok 2 - Test libssl.num
ok
test\recipes\01-test_symbol_presence.t ..
1..4
# NOTE: developper test!  It's possible that it won't run on your
# platform, and that's perfectly fine.  This is mainly for developers
# on Unix to check that our shared libraries are consistent with the
# ordinals (util/*.num in the source tree), something that should be
# good enough a check for the other platforms as well.
ok 1 # skip Can't run 'nm -Pg ..\libcrypto.so' => 256...  ignoring
ok 2 # skip Can't run 'nm -Pg ..\libcrypto.so' => 256...  ignoring
ok 3 # skip Can't run 'nm -Pg ..\libssl.so' => 256...  ignoring
ok 4 # skip Can't run 'nm -Pg ..\libssl.so' => 256...  ignoring
ok
test\recipes\05-test_bf.t ...............
1..1
testing blowfish in raw ecb mode
bftest.exe => 0
testing blowfish in ecb mode
testing blowfish set_key
testing blowfish in cbc mode
testing blowfish in cfb64 mode
testing blowfish in ofb64
ok 1 - running bftest
ok
test\recipes\05-test_cast.t .............
1..1
ecb cast5 ok
This test will take some time....123456789ABCDEF okcasttest.exe => 0

ok 1 - running casttest
ok
test\recipes\05-test_des.t ..............
1..1
Doing ecb
Doing ede ecb
destest.exe => 0
Doing cbc
Doing desx cbc
Doing ede cbc
Doing pcbc
Doing cfb8 cfb16 cfb32 cfb48 cfb64 cfb64() ede_cfb64() done
Doing ofb
Doing ofb64
Doing ede_ofb64
Doing cbc_cksum
Doing quad_cksum
input word alignment test 0 1 2 3
output word alignment test 0 1 2 3
fast crypt test
ok 1 - running destest
ok
test\recipes\05-test_hmac.t .............
1..1
test 0 ok
test 1 ok
hmactest.exe => 0
test 2 ok
test 3 ok
test 4 ok
test 5 ok
test 6 ok
ok 1 - running hmactest
ok
test\recipes\05-test_idea.t .............
1..1
ecb idea ok
cbc idea ok
ideatest.exe => 0
cfb64 idea ok
ok 1 - running ideatest
ok
test\recipes\05-test_md2.t ..............
1..0 # SKIP md2 is not supported by this OpenSSL build
skipped: md2 is not supported by this OpenSSL build
test\recipes\05-test_md4.t ..............
1..1
test 1 ok
test 2 ok
test 3 okmd4test.exe => 0

test 4 ok
test 5 ok
test 6 ok
test 7 ok
ok 1 - running md4test
ok
test\recipes\05-test_md5.t ..............
1..1
test 1 ok
test 2 ok
test 3 ok
md5test.exe => 0
test 4 ok
test 5 ok
test 6 ok
test 7 ok
ok 1 - running md5test
ok
test\recipes\05-test_mdc2.t .............
1..1
pad1 - ok
pad2 - okmdc2test.exe => 0

ok 1 - running mdc2test
ok
test\recipes\05-test_rand.t .............
1..1
Can't spawn "randtest.exe": No error at test\testlib/OpenSSL/Test.pm line 364.
randtest.exe => 255

not ok 1 - running randtest
#   Failed test 'running randtest'
#   at test\testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
test\recipes\05-test_rc2.t ..............
1..1
ecb RC2 ok
rc2test.exe => 0
ok 1 - running rc2test
ok
test\recipes\05-test_rc4.t ..............
1..1
test 0 ok
test 1 ok
test 2 okrc4test.exe => 0

test 3 ok
test 4 ok
test 5 ok
test end processing ....................done
test multi-call ....................done
bulk test ok
ok 1 - running rc4test
ok
test\recipes\05-test_rc5.t ..............
1..0 # SKIP rc5 is not supported by this OpenSSL build
skipped: rc5 is not supported by this OpenSSL build
test\recipes\05-test_rmd.t ..............
1..1
test 1 ok
test 2 ok
test 3 okrmdtest.exe => 0

test 4 ok
test 5 ok
test 6 ok
test 7 ok
test 8 ok
ok 1 - running rmdtest
ok
test\recipes\05-test_sha1.t .............
1..1
test 1 ok
test 2 ok
sha1test.exe => 0
test 3 ok
ok 1 - running sha1test
ok
test\recipes\05-test_sha256.t ...........
1..1
Testing SHA-256 ... passed.
Testing SHA-224 ... passed.
sha256t.exe => 0
ok 1 - running sha256t
ok
test\recipes\05-test_sha512.t ...........
1..1
Testing SHA-512 ... passed.
Testing SHA-384 ... passed.
sha512t.exe => 0
ok 1 - running sha512t
ok
test\recipes\05-test_wp.t ...............
1..1
Testing Whirlpool ......... passed.
wp_test.exe => 0
ok 1 - running wp_test
ok
test\recipes\10-test_bn.t ...............
1..3
ok 1 - require '..\test\recipes\bc.pl';
Can't spawn "cmd.exe": No error at test\testlib/OpenSSL/Test.pm line 364.
bntest.exe > tmp.bntest => 255

not ok 2 - initialize#   Failed test 'initialize'

#   at test\recipes\10-test_bn.t line 24.
# Looks like you failed 1 test of 3.
ok 3 # skip Initializing failed, skipping
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/3 subtests
        (less 1 skipped subtest: 1 okay)
test\recipes\10-test_exp.t ..............
1..1
Can't spawn "exptest.exe": No error at test\testlib/OpenSSL/Test.pm line 364.
exptest.exe => 255

not ok 1 - running exptest#   Failed test 'running exptest'

#   at test\testlib/OpenSSL/Test/Simple.pm line 77.
# Looks like you failed 1 test of 1.
Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests
test\recipes\15-test_dh.t ...............
1..1
Terminating on signal SIGINT(2)
Terminating on signal SIGINT(2)
NMAKE : fatal error UNMAKE : 10fatal error58:  Uterminated by user
Stop.
107
```
</summary>

rand, bn, exp, and dh failed before I terminated the process. Also worth noting that every time a test fails, windows pops up with the "searching for a solution thing" before I have to manually tell it to close the program and continue to the next test. Just in case that's relevent.

**Comment by alexander-novo** [comment]
I decided to open it up in VS debugger just to see if it could shed some light.

![image](https://cloud.githubusercontent.com/assets/11657716/15373931/bb354adc-1cfc-11e6-8d03-4ff2881db5fa.png)

It seems to fail at line 97 of threads_win.c

```
int CRYPTO_THREAD_run_once(CRYPTO_ONCE *once, void (*init)(void))
{
->  if (InitOnceExecuteOnce(once, once_cb, init, NULL))
        return 1;

    return 0;
}
```

**Comment by levitte** [comment]
@mattcaswell might be able to shed some light on the thread stuff...

I've a question, btw...  what perl dist are you using?

**Comment by alexander-novo** [comment]
ActiveState

**Comment by mattcaswell** [comment]
Hmmm....nothing immediately springing to mind. @ghedo any ideas?

**Comment by ghedo** [comment]
No idea, but it would probably be useful to call `GetLastError` after the failure. Hopefully it will have some useful information.

**Comment by mattcaswell** [comment]
This seems to be a 64 bit thing. I just "stole" my wife's laptop which has Windows 10 on it and installed VisualStudio on it (she's going to be mad at me). Everything works fine in 32 bit, but I get the same crash as above in 64 bit.

**Comment by mattcaswell** [comment]
This is *possibly* a compiler bug related to using the InitOnceExecuteOnce function. As far as I can tell we are doing things right, and this works in Win 8 32 bit and 64 bit, and Win 10 32 bit, but fails in Win 10 64 bit. A temporary workaround is to do the following when running Configure:

````
perl Configure VC-WIN64A -D_WIN32_WINNT=0x0501
````

**Comment by ghedo** [comment]
Oh, I just realized this is about a crash, not a function failure. I guess the suggestion to add GetLastError wasn't very useful....

**Comment by mattcaswell** [comment]
I figured this out. I spoke too soon BTW, this does impact Win 8 64 bit too.

The problem is this: threads_win.c just includes crypto.h which in turn includes windows.h. windows.h defines _WIN32_WINNT to an appropriate value for your OS (0x0A00 for win 10). Then we hit this bit in crypto.h:

````
#  if _WIN32_WINNT < 0x0600
typedef LONG CRYPTO_ONCE;
#   define CRYPTO_ONCE_STATIC_INIT 0
#  else
typedef INIT_ONCE CRYPTO_ONCE;
#   define CRYPTO_ONCE_STATIC_INIT INIT_ONCE_STATIC_INIT
#  endif
````
So `CRYPTO_ONCE` is `INIT_ONCE`.

When we compile init.c, it includes cryptlib_int.h, which in turn includes cryptlib.h which in turn includes e_os.h and crypto.h in that order. In e_os.h we have this:

````
#    define _WIN32_WINNT 0x0501
````
So, then when we subsequently include crypto.h we have `CRYPTO_ONCE` as a `LONG`!!!!! Hence the crash due to the difference in type definition between the two files.

The problem was introduced by commit 5c4328f04. That commit is problematic not only because of this issue, but also because it is using _WIN32_WINNT in a public header file to conditionally define things. Unfortunately _WIN32_WINNT maybe defined differently between OpenSSL build time and application build time.

**Comment by richsalz** [comment]
> Unfortunately _WIN32_WINNT maybe defined differently between OpenSSL build time and application build time.

Ouch, is that something that is generally supported?

Nice debugging.

**Comment by mattcaswell** [comment]
Yes. Unfortunately I do think we need to be able to support that scenario. :-(

**Comment by mattcaswell** [comment]
Should be fixed by commit fcb318c. Closing.

**Comment by kaduk** [comment]
> The simplest solution to this mess is just to always use the WinXP version
of CRYPTO_THREAD_run_once(). Its perhaps slightly sub-optimal but probably
not noticably.

So ... WinXP is not supported by Microsoft anymore.  Even Chrome has desupported XP by now.  Does OpenSSL plan to continue to support WinXP indefinitely, or is there some desupport timeline planned?
